### PR TITLE
Implement `pitch` argument for `pygame.image.frombuffer`, `pygame.image.fromstring`, and `pygame.image.frombytes`

### PR DESCRIPTION
**Pull request by avaxar**
Supposed to close pygame/pygame#3630 if I had the access. This PR implements a ~~`stride`~~ `pitch` argument for the functions `pygame.image.frombuffer`, `pygame.image.fromstring`, and `pygame.image.frombytes`. The argument is placed at the very back, so that it wouldn't ruin the behavior of existing code.

# Examples

```py
>>> pygame.image.fromstring(b"\xFF\xFF..\xFF\xFF..", (2, 2), "P", False, 4)
<Surface(2x2x8 SW)>
>>> pygame.image.frombytes(b"\xFF\xFF..\xFF\xFF..", (2, 2), "P", False, 4)
<Surface(2x2x8 SW)>
>>> pygame.image.frombuffer(b"\xFF\xFF..\xFF\xFF..", (2, 2), "P", 4)
<Surface(2x2x8 SW)>
```

```py
>>> pygame.image.fromstring(b"\xFF\xFF\xFF...\xFF\xFF\xFF...", (1, 2), "RGB", False, 6)
<Surface(1x2x24 SW)>
>>> pygame.image.frombytes(b"\xFF\xFF\xFF...\xFF\xFF\xFF...", (1, 2), "RGB", False, 6)
<Surface(1x2x24 SW)>
>>> pygame.image.frombuffer(b"\xFF\xFF\xFF...\xFF\xFF\xFF...", (1, 2), "RGB", 6)
<Surface(1x2x24 SW)>
```

```py
>>> pygame.image.fromstring(b"\xFF\xFF\xFF\xFF...\xFF\xFF\xFF\xFF...", (1, 2), "RGBA", False, 7)
<Surface(1x2x32 SW)>
>>> pygame.image.frombytes(b"\xFF\xFF\xFF\xFF...\xFF\xFF\xFF\xFF...", (1, 2), "RGBA", False, 7)
<Surface(1x2x32 SW)>
>>> pygame.image.frombuffer(b"\xFF\xFF\xFF\xFF...\xFF\xFF\xFF\xFF...", (1, 2), "RGBA", 7)
<Surface(1x2x32 SW)>
```

### CODE REVIEW

**Comment by avaxar** [comment]
I had to force push, because my commits from another PR got in. I shouldn't have done that PR in the `main` branch, I guess.

File: src_c/image.c [hunk 1]
```diff
@@ -1036,8 +1036,8 @@ image_frombytes(PyObject *self, PyObject *arg)
     __analysis_assume(format = "inited");
 #endif
 
-    if (!PyArg_ParseTuple(arg, "O!(ii)s|i", &PyBytes_Type, &bytes, &w, &h,
-                          &format, &flipped))
+    if (!PyArg_ParseTuple(arg, "O!(ii)s|ii", &PyBytes_Type, &bytes, &w, &h,
```
Comment (yunline) [inline src_c/image.c:1039]: `PyArg_ParseTupleAndKeywords` may be better here.

**Review by yunline — COMMENTED** [review]
👍LGTM

File: src_c/image.c (hunk 1 above)
Comment (avaxar) [reply src_c/image.c]: Implemented.

File: src_c/image.c [hunk 2]
```diff
@@ -1036,8 +1036,11 @@ image_frombytes(PyObject *self, PyObject *arg)
     __analysis_assume(format = "inited");
 #endif
 
-    if (!PyArg_ParseTuple(arg, "O!(ii)s|i", &PyBytes_Type, &bytes, &w, &h,
-                          &format, &flipped))
+    const char *kwids[] = {"bytes",   "size",   "format",
```
Comment (yunline) [inline src_c/image.c:1039]: `static` instead of `const`

File: src_c/image.c [hunk 3]
```diff
@@ -1208,7 +1258,9 @@ image_frombuffer(PyObject *self, PyObject *arg)
     __analysis_assume(format = "inited");
 #endif
 
-    if (!PyArg_ParseTuple(arg, "O(ii)s|i", &buffer, &w, &h, &format))
+    const char *kwids[] = {"buffer", "size", "format", "stride", NULL};
```
Comment (yunline) [inline src_c/image.c:1261]: `static` instead of `const`

**Comment by avaxar** [comment]
I was off for some time. Yeah, I've read the error messages. Will push the fix soon.

File: src_c/image.c [hunk 4]
```diff
@@ -1589,9 +1695,12 @@ static PyMethodDef _image_methods[] = {
 
     {"tostring", image_tobytes, METH_VARARGS, DOC_PYGAMEIMAGETOSTRING},
     {"tobytes", image_tobytes, METH_VARARGS, DOC_PYGAMEIMAGETOBYTES},
-    {"fromstring", image_frombytes, METH_VARARGS, DOC_PYGAMEIMAGEFROMSTRING},
-    {"frombytes", image_frombytes, METH_VARARGS, DOC_PYGAMEIMAGEFROMBYTES},
-    {"frombuffer", image_frombuffer, METH_VARARGS, DOC_PYGAMEIMAGEFROMBUFFER},
+    {"fromstring", image_frombytes, METH_VARARGS | METH_KEYWORDS,
```
Comment (yunline) [inline src_c/image.c:1698]: ~~image_tobytes~~  (PyCFunction)image_tobytes

File: src_c/image.c [hunk 5]
```diff
@@ -1589,9 +1695,12 @@ static PyMethodDef _image_methods[] = {
...
+     DOC_PYGAMEIMAGEFROMSTRING},
+    {"frombytes", image_frombytes, METH_VARARGS | METH_KEYWORDS,
```
Comment (yunline) [inline src_c/image.c:1700]: ~~image_frombytes~~ (PyCFunction)image_frombytes

File: src_c/image.c [hunk 6]
```diff
@@ -1589,9 +1695,12 @@ static PyMethodDef _image_methods[] = {
...
+     DOC_PYGAMEIMAGEFROMBYTES},
+    {"frombuffer", image_frombuffer, METH_VARARGS | METH_KEYWORDS,
```
Comment (yunline) [inline src_c/image.c:1702]: ~~image_frombuffer~~ (PyCFunction)image_frombuffer

File: src_c/image.c (hunk 4 above)
Comment (avaxar) [reply src_c/image.c:1698]: The other exported functions don't use the cast?

Comment (yunline) [reply src_c/image.c:1698]: Sorry i wrote it wrong. It's `image_frombytes`

**Comment by Starbuck5** [comment]
Hey, I haven't taken much of a look at this yet, but it seems cool.

You asked me in discord a while back if we should say `stride` or `pitch`, and I wasn't sure. I now have checked the docs and I see `Surface.get_pitch()` -- so to match this the terminology should be `pitch`.

**Review by MyreMylar — COMMENTED** [review]
I think this could use a test added to the `image_test.p` file demonstrating the usage to solve the problem outlined in the original issue:

> If you take a pygame surface, turn it into a buffer, then try to pass the buffer into pygame.image.frombuffer, it will most likely fail. Because most Surfaces have a pitch/stride that isn't equal to the width of the row.

Just need an awkwardly width'd surface and some calls. Then we can test it fails in old pygame and passes after this PR.

**Comment by avaxar** [comment]
Did that to rerun the CI.

**Comment by avaxar** [comment]
Ah, most of the CI servers' connection to install dependencies and such are dead.

**Comment by Starbuck5** [comment]
This looks great! Gave it some tests locally.

Before merge, I'd like to see a description of the pitch parameter added to the docs, as well as a "new in pygame 2.1.4" both for the parameter and the keyword arguments support.

If you rebase against main or merge main into your branch, you'll be able to grab the fix for the cppcheck option, which would be good.

**Review by Matiiss — COMMENTED** [review]
Also, more of a general thing but maybe the default values should be None instead of -1? None seems more Pythonic IMO (although it might be slower)

File: src_c/image.c [hunk 7]
```diff
@@ -1111,12 +1143,21 @@ image_frombytes(PyObject *self, PyObject *arg)
             Uint32 *pix = (Uint32 *)DATAROW(surf->pixels, looph, surf->pitch,
                                             h, flipped);
             memcpy(pix, data, w * sizeof(Uint32));
-            data += w * sizeof(Uint32);
+            data += pitch;
```
[the remark this answers is not in the document]
Comment (avaxar) [reply src_c/image.c:1146]: No. The pitch is in bytes.

File: src_c/image.c [hunk 8]
```diff
@@ -1047,7 +1050,16 @@ image_frombytes(PyObject *self, PyObject *arg)
...
 
     if (!strcmp(format, "P")) {
-        if (len != (Py_ssize_t)w * h)
+        if (pitch == -1) {
+            pitch = w;
+        }
+        else if (pitch < w) {
+            return RAISE(PyExc_ValueError,
+                         "Pitch must be greater than or equal to the width "
+                         "by the format");
```
[the remark this answers is not in the document]
Comment (avaxar) [reply src_c/image.c:1059]: No. What I meant was that, the stride/pitch should be greater than or equal to what a horizontal line in the surface takes in bytes as defined **by the format** ("P", `width` bytes;  "RGB", `width * 3` bytes; "RGBA"z `width * 4`; and so on).

[the remark this answers is not in the document]
Comment (Matiiss) [reply src_c/image.c:1059]: Then maybe "as per the format"? I don't quite see how "by" fits here properly

[the remark this answers is not in the document]
Comment (Starbuck5) [reply src_c/image.c:1059]: I agree with Matt, however I want to keep this PR moving, so I may approve/merge without this changed.

**Review by Starbuck5 — APPROVED** [review]

File: src_c/image.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (avaxar) [reply src_c/image.c]: Changed accordingly.

**Review by Matiiss — APPROVED** [review]
LGTM

**Pull request merged**